_posts/2020-03-06-introducing-mirrormaker2.md

```diff
@@ -0,0 +1,268 @@
+  topicsPattern: ".*"
+```
+
+You use `topicsBlacklistPattern` if you want to use blacklists.
+
+Pretty similar to before, but previously topic _configuration_ was not replicated.
+The target cluster had to be manually maintained to match topic configuration changes in the source cluster.
+
+This time the topic configuration is automatically synchronized between source and target clusters through configuration properties.
+Configuration changes are propagated to remote topics so that new topics and partitions are detected and created.
+The need for rebalancing vanishes.
+
+### Offset tracking and mapping
+
+In the old version of MirrorMaker, the offset of the source topic in the target cluster begins when the replication begins.
+The `__consumer_offsets` topic is not mirrored.
+So offsets of the source topic and its replicated equivalent can have two entirely different positions.
+This can be problematic in a failover situation.
+How to find the offset in the target cluster?
+Strategies such as using timestamps can be adopted, but it adds complexity.
+
+Poof! These issues entirely disappear with MirrorMaker 2.0.
+Instead, we get simplicity and sophistication through the `MirrorCheckpointConnector`.
+
+`MirrorCheckpointConnector` tracks and maps offsets for specified consumer groups using an _offset sync_ topic and _checkpoint_ topic.
+The _offset sync_ topic maps the source and target offsets for replicated topic partitions from record metadata.
+A _checkpoint_ is emitted from each source cluster and replicated in the target cluster through the _checkpoint_ topic.
+The _checkpoint_ topic maps the last committed offset in the source and target cluster for replicated topic partitions in each consumer group.
+
+If you want automatic failover, you can add Kafka's new `RemoteClusterUtils.java` utility class to your consumers.
+The class translates the consumer group offset from the source cluster to the corresponding offset for the target cluster.
+
+The consumer groups tracked by `MirrorCheckpointConnector` are dependent on those defined in a _whitelist_ or _blacklist_:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1alpha1
+kind: KafkaMirrorMaker2
+metadata:
+  name: my-mirror-maker2
+spec:
+  # ...
+  groupsPattern: "group1|group2|group3"
+```
+
+You use `groupsBlacklistPattern` if you want to use blacklists.
+
+### Checking the connection
+
+The old way of checking that MirrorMaker was working was by using standard Kubernetes _healthcheck_ probes to know when MirrorMaker can accept traffic and when it needs a restart.
+MirrorMaker 2.0 periodically checks on the connection through its dedicated `MirrorHeartbeatConnector`.
+
+`MirrorHeartbeatConnector` periodically checks connectivity between clusters using the _heartbeat_ topic.
+A _heartbeat_ is emitted from each source cluster and replicated in the target cluster through the _heartbeat_ topic.
+The _heartbeat_ topic is used to check that the source cluster is available and the clusters are connected.
```

Comment:
I have a more technical question here because the paragraph is not clear to me.
If I have just a source and a target cluster, MM2 is running on the target one. It's based on Kafka Connect and works in pull mode, pulling messages from the source; it sounds to me that the source cluster doesn't know that "someone" is mirroring the messages, it doesn't know about MM2 on the other side. So how can it produce a heartbeat message into a local heartbeat topic? Tbh it seems to be what the KIP-382 says (https://cwiki.apache.org/confluence/display/KAFKA/KIP-382%3A+MirrorMaker+2.0) but in one of my deployments I just see a `heartbeats` topic on the target cluster where MM2 is running and nothing related to heartbeat on the source cluster. @ajborley  maybe a question for you?

Reply:
Thanks Paolo. @ajborley - heartbeats and source clusters. Is there anything we can add to clarify what happens to produce a heartbeat at the source?

Reply:
The heartbeat topic and records are created on the target cluster (the cluster that is local to MM2/Kafka Connect), not the source cluster (remote). The idea is to have MM2 clusters on both sides, so that a heartbeat produced locally by the HeartbeatConnector running at the source cluster is treated as a normal remote topic to be mirrored by the MM2 SourceConnector running at the target.
 
By mirroring the heartbeat topics, it’s easy to check that the connection between source and target is stable and if things do go wrong, the offset position and record time stamps can help with recovery and diagnosis.

Hope that makes some kind of sense!

Reply:
Updated to:

> `MirrorHeartbeatConnector` periodically checks connectivity between clusters using the _heartbeat_ topic.
> A _heartbeat_ is emitted from each remote (source) cluster and mirrored to the target cluster by the `MirrorSourceConnector` as a _heartbeat_ topic.
> The _heartbeat_ topic makes it easy to check that the source cluster is available and the clusters are connected.
> If things go wrong, the _heartbeat_ topic offset positions and time stamps can help with recovery and diagnosis.

Reply:
@PaulRMellor but it doesn't explain the scenario where you have just one MM2 on the target cluster. So we should be clear that what you wrote makes sense when you have MM2 on both source and target otherwise it doesn't. 

@ajborley reading your explanation (thanks for that!) sounds to me that when you have a scenario with MM2 just running on the target, the _hearthbeat_ topic is quite useless, is that right?

Reply:
@ppatierno - agreed, the heartbeat topic doesn’t have a lot of utility when it’s not being mirrored. It does show you that your MM2 cluster can continuously connect and produce to the local Kafka, which could perhaps be useful, particularly when the topics that are being mirrored to the local Kafka (via the MM2 MirrorSourceConnector) are bursty or sporadic.

Reply:
@ajborley agreed. So @PaulRMellor I think we should reflect this much better in the above paragraph.

Reply:
How about this:

> `MirrorHeartbeatConnector` periodically checks connectivity between clusters.
> A _heartbeat_ is emitted from each remote (source) cluster and a _heartbeat_ topic is created on a target cluster. 
> If you have MirrorMaker 2.0 at both the remote and target location, a _heartbeat_ emitted locally by the `MirrorHeartbeatConnector` is treated like any remote topic and mirrored by the `MirrorSourceConnector` at the target cluster.
> The _heartbeat_ topic makes it easy to check that the source cluster is available and the clusters are connected.
> If things go wrong, the _heartbeat_ topic offset positions and time stamps can help with recovery and diagnosis.

Reply:
> A _heartbeat_ is emitted from each remote (source) cluster and a heartbeat topic is created on a target cluster.

This isn't quite right - how about:

 A _heartbeat_ record is produced every second by the `MirrorHeartbeatConnector` into a _heartbeat_ topic that is created on the target cluster.

Reply:
Well, source and target clusters naming starting to be not well defined.
For example, consider the case where I have a source cluster (where producers send messages) and a target cluster (where consumers get mirrored messages). I have MM2 on the target cluster for mirroring but we decide to leverage the heartbeat feature so we need to install MM2 on the source cluster as well. In any case, the source cluster doesn't want to mirror any other topics from the target but just the heartbeat right? In this case, a heartbeat topic is created on the source cluster as well and mirrored to the target, isn't it?

Reply:
Thanks @ajborley. Latest:

> `MirrorHeartbeatConnector` periodically checks connectivity between clusters.
> A _heartbeat_ is produced every second by the `MirrorHeartbeatConnector` into a _heartbeat_ topic that is created on the target cluster.
> If you have MirrorMaker 2.0 at both the remote and target location, the _heartbeat_ emitted locally by the `MirrorHeartbeatConnector` is treated like any remote topic and mirrored by the `MirrorSourceConnector` at the target cluster.
> The _heartbeat_ topic makes it easy to check that the source cluster is available and the clusters are connected.
> If things go wrong, the _heartbeat_ topic offset positions and time stamps can help with recovery and diagnosis.

Reply:
@ppatierno - you're right. In your case the MM2 running at the source cluster is only running the `MirrorHeartbeatConnector`, and the MM2 running at the target cluster is only running the `MirrorSourceConnector`, causing the `heartbeat` topic to be created on the source cluster and mirrored to the target cluster.

Perhaps we should change this to use the terms `local` and `remote`? How does the following sound:

> `MirrorHeartbeatConnector` periodically checks connectivity between clusters.
A _heartbeat_ is produced every second by the `MirrorHeartbeatConnector` into a _heartbeat_ topic that is created on the local cluster.
If you have MirrorMaker 2.0 at both the remote and local locations, the _heartbeat_ emitted at the remote location by the `MirrorHeartbeatConnector` is treated like any remote topic and mirrored by the `MirrorSourceConnector` at the local cluster.
The _heartbeat_ topic makes it easy to check that the remote cluster is available and the clusters are connected.
If things go wrong, the _heartbeat_ topic offset positions and time stamps can help with recovery and diagnosis.

Reply:
@ajborley yeah I like it. @PaulRMellor wdyt?

Reply:
Good for me :+1:

Reply:
@ppatierno@ajborley. Thanks. I've updated as suggested.

Reply:
we are good to merge than!